[IsDeprecatedWeakRefSmartPointerException] Make more WebAuthentication classes RefCounted or equivalent

basuke opened the pull request:
#### baf68af71e1f9fb6f560653be5d2185b053d60bd
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make more WebAuthentication classes RefCounted or equivalent
<a href="https://bugs.webkit.org/show_bug.cgi?id=281178">https://bugs.webkit.org/show_bug.cgi?id=281178</a>

Reviewed by Chris Dumez.

Make following classes refCounted or equivalent because they are subclass of CanMakeWeakPtr:
- Authenticator
- AuthenticatorManager
- AuthenticatorPresenterCoordinator
- WebAuthenticationPanelClient

Other classes for WebAuthentication are treated in the next patch.

* Source/WebKit/UIProcess/API/APIWebAuthenticationPanel.cpp:
(API::WebAuthenticationPanel::WebAuthenticationPanel):
(API::WebAuthenticationPanel::setClient):
* Source/WebKit/UIProcess/API/APIWebAuthenticationPanel.h:
* Source/WebKit/UIProcess/API/APIWebAuthenticationPanelClient.h:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(-[_WKWebAuthenticationPanel delegate]):
(-[_WKWebAuthenticationPanel setDelegate:]):
* Source/WebKit/UIProcess/WebAuthentication/Authenticator.cpp:
(WebKit::Authenticator::receiveRespond const):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::AuthenticatorManager::serviceStatusUpdated):
(WebKit::AuthenticatorManager::authenticatorStatusUpdated):
(WebKit::AuthenticatorManager::requestPin):
(WebKit::AuthenticatorManager::selectAssertionResponse):
(WebKit::AuthenticatorManager::requestLAContextForUserVerification):
(WebKit::AuthenticatorManager::runPresenterInternal):
(WebKit::AuthenticatorManager::invokePendingCompletionHandler):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.mm:
(WebKit::AuthenticatorPresenterCoordinator::create):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticationPanelClient.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticationPanelClient.mm:
(WebKit::WebAuthenticationPanelClient::create):
(WebKit::WebAuthenticationPanelClient::WebAuthenticationPanelClient):
(WebKit::WebAuthenticationPanelClient::delegate const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::protectedAuthenticatorManager):
(WebKit::WebsiteDataStore::protectedVirtualAuthenticatorManager):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: <a href="https://commits.webkit.org/285204@main">https://commits.webkit.org/285204@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e99e48b1c39822acf5ddc6aa32da8942e209139a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71841 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51254 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24615 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23046 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59055 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22866 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/75999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15221 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74907 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/46505 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61895 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43169 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21396 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65080 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19729 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77676 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16075 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/18921 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16119 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61918 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64443 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15881 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12616 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6262 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47054 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1844 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48125 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49409 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47867 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Authenticator.h:
```diff
@@ -50,6 +50,9 @@ using AuthenticatorObserverRespond = std::variant<Ref<WebCore::AuthenticatorResp
 public:
     virtual ~AuthenticatorObserver() = default;
+
+    DECLARE_VIRTUAL_REFCOUNTED;
```
The code was correct before. No reason to hide behind a macro.

basuke replied:
If macro exists, then we should use the macro for consistency.

cdumez replied:
I believe the plan is to move away from the macro and remove it.

basuke replied:
This macro was introduced by Geoff just three weeks ago. https://github.com/WebKit/WebKit/commit/5d100c24f90237a63172d016c390a3bd223293e6

cdumez replied:
I know and I've objected to this macro since its introduction. It was discussed during last Thursday Safer C++ Weekly sync.

basuke replied:
Oh, I see. Okay, I. revert this for now.

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.h:
```diff
@@ -72,6 +63,8 @@ class AuthenticatorManager : public RefCounted<AuthenticatorManager>, public Aut
     static Ref<AuthenticatorManager> create();
     virtual ~AuthenticatorManager() = default;
 
+    DEFINE_VIRTUAL_REFCOUNTED;
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.mm:
```diff
@@ -104,6 +104,21 @@
+
+void AuthenticatorPresenterCoordinator::ref() const
+{
+    protectedAuthenticatorManager()->ref();
```
Do we really need `protectedAuthenticatorManager()` just to call `ref()`?

basuke replied:
Right. Static Analyzer doesn't care about the protection of the object for calling ref/deref(). Will remove.

basuke replied:
This is WeakPtr so still checking is required before using it. Is this change correct for you?
```C++
void AuthenticatorPresenterCoordinator::ref() const
{
    if (RefPtr manager = m_manager.get())
        manager->ref();
}

void AuthenticatorPresenterCoordinator::deref() const
{
    if (RefPtr manager = m_manager.get())
        manager->deref();
}
```

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.mm:
```diff
@@ -104,6 +104,21 @@
+
+void AuthenticatorPresenterCoordinator::deref() const
+{
+    protectedAuthenticatorManager()->deref();
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
```diff
@@ -508,7 +508,7 @@ void AuthenticatorManager::runPresenter()
 {
     auto& options = m_pendingRequestData.options;
-    m_presenter = makeUnique<AuthenticatorPresenterCoordinator>(*this, getRpId(options), transports, getClientDataType(options), getUserName(options));
+    m_presenter = makeUniqueWithoutRefCountedCheck<AuthenticatorPresenterCoordinator>(*this, getRpId(options), transports, getClientDataType(options), getUserName(options));
```
This cannot be correct. m_presenter gets nulled out in `AuthenticatorManager::clearState()`, therefore `AuthenticatorPresenterCoordinator` cannot forward its ref-counting to `AuthenticatorManager`.

basuke replied:
Thanks. Yeah, I totally misunderstood the lifecycle. Changed to RefCountedAndCanMakeWeakPtr.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -326,6 +326,8 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
     void setMockWebAuthenticationConfiguration(WebCore::MockWebAuthenticationConfiguration&&);
     VirtualAuthenticatorManager& virtualAuthenticatorManager();
+
+    Ref<AuthenticatorManager> protectedAuthenticatorManager();
```
We should move this above, right after `authenticatorManager()`

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticationPanelClient.mm:
```diff
@@ -53,7 +53,17 @@
-RetainPtr<id <_WKWebAuthenticationPanelDelegate> > WebAuthenticationPanelClient::delegate()
+void WebAuthenticationPanelClient::ref() const
+{
+    [delegate() retain];
```
This still doesn't seem right? As far as I can tell, `WebAuthenticationPanelClient` is owned by the `WebAuthenticationPanel` not the `_WKWebAuthenticationPanelDelegate`. Also, `WebAuthenticationPanel`'s client may get overwritten via `setClient()`.

cdumez commented on Source/WebKit/UIProcess/API/APIWebAuthenticationPanelClient.h:
```diff
@@ -43,9 +43,10 @@ class AuthenticatorAssertionResponse;
 namespace API {
 
-class WebAuthenticationPanelClient {
+class WebAuthenticationPanelClient : public RefCounted<WebAuthenticationPanelClient> {
```
Since this is an interface, @geoffreygaren has been recommending we provide a virtual ref() / deref() in such cases and let the subclasses decide how to do their ref-counting.

basuke replied:
Actually that is a concrete class which is used in `WebAuthenticationPanel` as the default implementation. This is bad design but it is required like that.

basuke replied:
We can make it as pure abstraction class and define local class for empty implementation.

cdumez replied:
> Actually that is a concrete class which is used in WebAuthenticationPanel as the default implementation. This is bad design but it is required like that.

In the future, if my review feedback is based on an incorrect assumption, please confirm with me before making further changes to avoid wasting time. I would have approved the PR as-is. Now the new PR is not as good.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -154,15 +154,16 @@ - (void)dealloc
 {
-    auto client = WTF::makeUniqueRef<WebKit::WebAuthenticationPanelClient>(self, delegate);
-    _client = client.get();
+    auto client = WebKit::WebAuthenticationPanelClient::create(self, delegate);
```
`auto` -> `Ref`

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticationPanelClient.h:
```diff
@@ -34,27 +34,23 @@
+    static Ref<WebAuthenticationPanelClient> create(_WKWebAuthenticationPanel *, id<_WKWebAuthenticationPanelDelegate>);
 
-    RetainPtr<id <_WKWebAuthenticationPanelDelegate> > delegate();
+    void ref() const { RefCounted::ref(); }
```
`final`

basuke replied:
If moving back to previous, then this cannot be final because ref() is not virtual in RefCounted.

basuke replied:
...then no need to define them!? Let's delete them.

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticationPanelClient.h:
```diff
@@ -34,27 +34,23 @@
 
-    RetainPtr<id <_WKWebAuthenticationPanelDelegate> > delegate();
+    void ref() const { RefCounted::ref(); }
+    void deref() const { RefCounted::deref(); }
```
`final`

basuke replied:
Ditto

cdumez commented on Source/WebKit/UIProcess/API/APIWebAuthenticationPanel.cpp:
```diff
@@ -37,14 +37,26 @@ namespace API {
 using namespace WebCore;
 using namespace WebKit;
 
+class NullPanelClient final : public RefCounted<NullPanelClient>, public WebAuthenticationPanelClient {
```
Why do we need this now? The new PR looks significantly worse than the earlier one. I think we should go back to the earlier iteration if someone was constructing a `WebAuthenticationPanelClient` (and not a subclass of `WebAuthenticationPanelClient`).

basuke replied:
I don't think this is a bad code. Interface should not be used as concrete class in my opinion so I agreed to change the code. But I don't have enough time to work with this code, I reverted this as you suggested.

cdumez commented on Source/WebKit/UIProcess/API/APIWebAuthenticationPanelClient.h:
```diff
@@ -43,9 +43,10 @@ class AuthenticatorAssertionResponse;
+class WebAuthenticationPanelClient : public RefCounted<WebAuthenticationPanelClient> {
     WTF_MAKE_TZONE_ALLOCATED_INLINE(WebAuthenticationPanelClient);
 public:
+    static Ref<WebAuthenticationPanelClient> create() { return adoptRef(*new WebAuthenticationPanelClient); }
```
We need to make the following constructor protected too, otherwise nothing is forcing clients to use `create()`:
```
protected:
    WebAuthenticationPanelClient() = default;
```

cdumez commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -2166,6 +2166,11 @@ void WKPageSetPageUIClient(WKPageRef pageRef, const WKPageUIClientBase* wkClient
             public:
+                static Ref<PanelClient> create()
+                {
+                    return adoptRef(*new PanelClient());
```
nit: we don't need `()`, you did it right above so it's not consistent.

cdumez commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -2176,14 +2181,17 @@ void WKPageSetPageUIClient(WKPageRef pageRef, const WKPageUIClientBase* wkClient
                 }
+
+            private:
+                PanelClient() = default;
```
Here you did it correctly.

cdumez reviewed (approved):
LGTM with nits.

basuke committed:
[IsDeprecatedWeakRefSmartPointerException] Make more WebAuthentication classes RefCounted or equivalent
https://bugs.webkit.org/show_bug.cgi?id=281178

Reviewed by Chris Dumez.

Make following classes refCounted or equivalent because they are subclass of CanMakeWeakPtr:
- Authenticator
- AuthenticatorManager
- AuthenticatorPresenterCoordinator
- WebAuthenticationPanelClient

Other classes for WebAuthentication are treated in the next patch.

* Source/WebKit/UIProcess/API/APIWebAuthenticationPanel.cpp:
(API::WebAuthenticationPanel::WebAuthenticationPanel):
(API::WebAuthenticationPanel::setClient):
* Source/WebKit/UIProcess/API/APIWebAuthenticationPanel.h:
* Source/WebKit/UIProcess/API/APIWebAuthenticationPanelClient.h:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(-[_WKWebAuthenticationPanel delegate]):
(-[_WKWebAuthenticationPanel setDelegate:]):
* Source/WebKit/UIProcess/WebAuthentication/Authenticator.cpp:
(WebKit::Authenticator::receiveRespond const):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::AuthenticatorManager::serviceStatusUpdated):
(WebKit::AuthenticatorManager::authenticatorStatusUpdated):
(WebKit::AuthenticatorManager::requestPin):
(WebKit::AuthenticatorManager::selectAssertionResponse):
(WebKit::AuthenticatorManager::requestLAContextForUserVerification):
(WebKit::AuthenticatorManager::runPresenterInternal):
(WebKit::AuthenticatorManager::invokePendingCompletionHandler):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.mm:
(WebKit::AuthenticatorPresenterCoordinator::create):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticationPanelClient.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticationPanelClient.mm:
(WebKit::WebAuthenticationPanelClient::create):
(WebKit::WebAuthenticationPanelClient::WebAuthenticationPanelClient):
(WebKit::WebAuthenticationPanelClient::delegate const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::protectedAuthenticatorManager):
(WebKit::WebsiteDataStore::protectedVirtualAuthenticatorManager):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: https://commits.webkit.org/285204@main
